rocAL/rocAL/include/meta_data/meta_data.h

```diff
@@ -153,78 +235,129 @@ struct MetaDataBatch
     }
-    virtual std::shared_ptr<MetaDataBatch> clone()  = 0;
-    std::vector<int>& get_label_batch() { return _label_id; }
-    std::vector<BoundingBoxCords>& get_bb_cords_batch() { return _bb_cords; }
-    std::vector<BoundingBoxCords_xcycwh>& get_bb_cords_batch_xcycxwh() { return _bb_cords_xcycwh; }
-    std::vector<BoundingBoxLabels>& get_bb_labels_batch() { return _bb_label_ids; }
-    ImgSizes & get_img_sizes_batch() { return _img_sizes; }
-    JointsDataBatch & get_joints_data_batch() { return _joints_data; }
+    virtual std::shared_ptr<MetaDataBatch> clone() = 0;
+    virtual void get_labels_batch(std::vector<Labels>** label_ids) {}
+    virtual void get_bb_cords_batch(std::vector<BoundingBoxCords>** bb_cords) {}
+    virtual void get_bb_cords_batch_xcycxwh(std::vector<BoundingBoxCords_xcycwh>** bb_cords_xcycwh) {}
+    virtual void get_mask_cords_batch(std::vector<MaskCords>** mask_cords) {}
+    virtual void get_mask_polygons_count_batch(std::vector<std::vector<int>>** polygon_counts) {}
+    virtual void get_mask_vertices_count_batch(std::vector<std::vector<std::vector<int>>>** vertices_count) {}
+    virtual void get_joints_data_batch(JointsDataBatch** joints_data) {}
+    std::vector<uint>& get_image_id_batch() { return _info_batch.img_ids; }
+    std::vector<std::string>& get_image_names_batch() {return _info_batch.img_names; }
+    ImgSizes& get_img_sizes_batch() { return _info_batch.img_sizes; }
+    MetaDataInfoBatch& get_info_batch() { return _info_batch; }
+    void set_metadata_type(MetaDataType metadata_type) { _type = metadata_type; }
+    MetaDataType get_metadata_type() { return _type; }
 protected:
-    std::vector<int> _label_id = {}; // For label use only
-    std::vector<BoundingBoxCords> _bb_cords = {};
-    std::vector<BoundingBoxCords_xcycwh> _bb_cords_xcycwh = {};
-    std::vector<BoundingBoxLabels> _bb_label_ids = {};
-    std::vector<ImgSize> _img_sizes = {};
-    JointsDataBatch _joints_data = {};
+    MetaDataInfoBatch _info_batch;
+    MetaDataType _type;
+};
+
+template <typename T>
+void (MetaDataBatch::*pGetMetaDataBatchFunction)(T **ptr);
+
+template <typename T>
+T& getMetaDataBatchValues(MetaDataBatch &metadatabatch, void(MetaDataBatch::*pGetMetaDataBatchFunction)(T **ptr)) {
+    T *val;
+    (metadatabatch.*pGetMetaDataBatchFunction)(&val);
+    return *val;
 };
 
 struct LabelBatch : public MetaDataBatch
 {
     void clear() override
     {
-        _label_id.clear();
+        for (auto label : _label_ids) {
+            label.clear();
+        }
+        _info_batch.clear();
+        _label_ids.clear();
+        _buffer_size.clear();
     }
     MetaDataBatch&  operator += (MetaDataBatch& other) override
     {
-        _label_id.insert(_label_id.end(), other.get_label_batch().begin(), other.get_label_batch().end());
+        _label_ids.insert(_label_ids.end(),
+                          getMetaDataBatchValues<std::vector<Labels>>(other, &MetaDataBatch::get_labels_batch).begin(),
```

Comment:
please don't use getMetaDataBatchValues function. 
Also you can't define an operator for base class inside a derived class. Here the operator has to be defined for LabelBatch ( not MetaDataBatch).
We need to go through the design of the metadata again. I think we are using inheritance not the right way and making it work through some hacks

Reply:
- This templated function was mainly introduced to fix warning that arises from changing all these functions to virtual in the base class.
- In previous implementation we had all the data members as a part of base class. hence the get_labels / get bbox API was able to directly return the values
- In current implementation, all data members are present in respective derived classed so we have changed all these to virtual functions with return type as void and added support to return the data by reference using templated function.
- In the getMetadataValues function. We use the function pointer and invoke the function of the derived class using the object of the derived class that is passed to this function.
- Here we accept the object with base class name so it can be applicable to all derived classes.

T& getMetaDataValues(MetaData &metadata, void(MetaData::*pGetMetaDataFunction)(T** ptr))
{
    T* val;
    (metadata.*pGetMetaDataFunction)(&val);
    return *val;
};